Update total model parameters computation, adds ability to define model label

JayOrten opened the pull request:
- Before, we were using Torchinfo to obtain the total parameters of the model before we begin training. With the addition of flash attention, it seems that this cause OOM errors on VRAM for very large models. This is likely a problem with Torchinfo. This replaces the Torchinfo code with simple PyTorch code that sums up all model parameters. Functionality remains the same.

- Before, we were creating a unique `model_label` for each run, which includes the time stamp. This is nice, but when the job is requeued through Slurm it creates a new folder and doesn't find the checkpoint, thus restarting the training from the beginning. Instead, the code should reference the same model folder. To facilitate this, I have added a new configuration parameter, `model_label`, which the use can define as the name of the folder to save run info(checkpoints, logs). In my experience, this approach is much easier to use and provides the user with more control. If this isn't supplied, a label will be generated. However, this may begin using an already existing folder, which is the biggest downside. So the user needs to be careful.

JayOrten committed:
Add numpy import and update model parameter calculation to use basic pytorch

JayOrten committed:
add model_name param

JayOrten committed:
add model_name logic, place logs within the model folder

JayOrten committed:
Update model_name to model_label in template_config.yaml

KimballNJardine commented on src/train_model.py:
```diff
@@ -71,7 +71,7 @@ def train_model(config: Struct):
     else:
         raise ValueError(f"Model type '{config.model_type}' not supported!")
-
+        
```
Remove extra spaces from empty line.

JayOrten replied:
done

KimballNJardine commented on src/train_model.py:
```diff
@@ -85,18 +85,18 @@ def train_model(config: Struct):
+    else:
+        model_label = f"{config.model_type}_{total_params}" + \
+            f"_LR{config.learning_rate}_ED{config.embed_dim}" + \
+            f"_FFN{config.ffn_dim}_H{config.heads}_S{config.seq_len}"
```
Do we want to remove the time at the end (`datetime.now().strftime('%Y-%m-%d-%H:%M:%S')`)?

JayOrten replied:
Yes, because that's what it making it unique. It's nice until the requeuing doesn't work, because model_label will have a unique timestamp everytime the script is run. This PR is essentially proposing that the model_label should preferably be defined in the config.

KimballNJardine reviewed (changes requested):
Minor change requested and question to resolve

DrewGalbraith committed:
switch np implementation to torch

I thought there was a way to do this in torch to avoid introducing another dependency and I found+adapted+tested [this alternative](https://saturncloud.io/blog/check-the-total-number-of-parameters-in-a-pytorch-model/#:~:text=total_params%20%3D%20sum(p.numel()%20for%20p%20in%20model.parameters())). (Plus, it's a little more concise).

JayOrten committed:
remove space

JayOrten commented on src/train_model.py:
```diff
@@ -86,8 +85,7 @@ def train_model(config: Struct):
     # Get number of parameters in model
-    model_parameters = filter(lambda p: p.requires_grad, model.parameters())
-    total_params = sum([np.prod(p.size()) for p in model_parameters])
+    total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
```
@DrewGalbraith , thanks, looks good

KimballNJardine reviewed (approved):
Looks great. Approved.

JayOrten committed:
Merge branch 'main' into fix-model-param
